### DISCUSSION TITLE
MenuItem command issue or bug?

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#13546] [Q&A] [answered]
Hi, I came accros an anoying issue with MenuItem command.

I was implementing a Recent Items submenu, where when you click on a recent item MenuItem, the corresponding file is opened and the recent item is moved at the top of the list.

When you click on the submenu item, no problem, the file is opened and the menu is closed.

However, when you click on another item than the first, the command is executed in the pointer down, then the recent item is moved a the top of the list, as well as the menu item which is binding using ItemsSource. The problem is that the pointer up is now executed on another item (as the clicked item position changed), and so the menu is not closed after click... 

It is probably a bug that the command is executed in the pressed event, and not in the released event (only if the pointer is released on same item than the clicked item). I think it should be executed in the same place as the code that close the menu.

Please let me know if I should open a new bug or not? I think there are already other bugs related to command executed on pressed, instead of released event.

**Comment by stevemonaco** [comment, marked as answer]
I'm not sure if this is a bug. As a workaround, I `await Task.Yield();` in these cases. I also do it for context menus. Alternatively, you might be able to post it to the dispatcher the work takes place after the menu is closed.

```cs
private async void Handle(ProjectLoadedMessage message) // async void here for Messenger compatibility, don't mind
{
    await Task.Yield(); // Yield so that the menu can close, otherwise changing the collection keeps it open

    if (RecentProjectFiles.Contains(message.ProjectFileName))
    {
        RecentProjectFiles.Remove(message.ProjectFileName);
        RecentProjectFiles.Insert(0, message.ProjectFileName);
    }
    else
    {
        RecentProjectFiles.Insert(0, message.ProjectFileName);
        if (RecentProjectFiles.Count > 8)
            RecentProjectFiles = new(RecentProjectFiles.Take(8));
    }
}
```

**Reply by dbriard** [reply]
@stevemonaco Thanks a lot for the workaround, it work perfectly!
But that's still strange to me that click that close the menu, and the command are not executed at the same place.

**Reply by StefanKoell** [reply]
I had something like this in WinForms where I handled the mouse down and mouse up event where the down event changed the UI and the up event was then fired on the new and replaced UI. It was very confusing to track down but the solution was quite the same: waiting for the event loop to continue and replace the UI at the end (after the mouse up event was processed).

Maybe a similar scenario.
